lang/src/org/partiql/lang/eval/io/DelimitedValues.kt

```diff
@@ -131,34 +124,36 @@ object DelimitedValues {
     fun writeTo(ion: IonSystem,
                 output: Writer,
                 value: ExprValue,
-                delimiter: String,
+                delimiter: Char,
                 newline: String,
-                writeHeader: Boolean): Unit {
-        val nullValue = ion.newNull()
+                writeHeader: Boolean) {
+        val csvPrinter = CSVPrinter(output, CSVFormat.DEFAULT.withDelimiter(delimiter).withRecordSeparator(newline))
         var names: List<String>? = null
         for (row in value) {
             val colNames = row.orderedNames
                 ?: throw IllegalArgumentException("Delimited data must be ordered tuple: $row")
             if (names == null) {
                 // first row defines column names
                 names = colNames
-
                 if (writeHeader) {
-                    names.joinTo(output, delimiter)
-                    output.write(newline)
+                    csvPrinter.printRecord(names)
                 }
-            } else if (names != colNames) {
-                // mismatch on the tuples
+            } else if (names != colNames) { // We need to check if the column names in other rows are all the same as the first one's.
                 throw IllegalArgumentException(
                     "Inconsistent row names: $colNames != $names"
                 )
             }
 
-            names.map {
-                val col = row.bindings[BindingName(it, BindingCase.SENSITIVE)]?.ionValue ?: nullValue
-                col.csvStringValue()
-            }.joinTo(output, delimiter)
-            output.write(newline)
+            csvPrinter.printRecord(
+                names.map {
+                    val col = row.bindings[BindingName(it, BindingCase.SENSITIVE)]?.ionValue ?: ion.newNull()
+                    col.csvStringValue()
+                }
+            )
+
+            csvPrinter.flush()
```

Comment:
Why call `.flush()` here?  Seems like that could hurt performance since you're forcing more smaller writes to disk instead of allowing the OS to batch the writes together.

Reply:
Will remove it in the next revision.